File: docs/palindromes-in-peg/scaffold_circuit_search.py
"""Predicated lowering of the match-paced DP search used by scaffold_galil."""
from fpp_finite import LEFT, END
from fpp_subroutine import SOURCE
from dp_finite import LOWER
from scaffold_circuit import (Value, PREVIOUS, TRUE, FALSE, choose,
                              conjunction as AND, disjunction as OR, neg as NOT)
from scaffold_circuit_structs import Counter


class Search:
  MODES = ("idle", "grow", "lower", "lower_home", "copy", "home", "run",
           "found", "missed", "wait", "double")

  def __init__(self, circuit, program, pool, center, walker, radius):
    self.circuit, self.program = circuit, program
    self.center, self.walker, self.radius = center, walker, radius
    self.lower, self.span, self.work, self.debt = [Counter(pool, "sp." + name)
                                                for name in ("lo", "span", "work", "debt")]
    self.mode = circuit.get(PREVIOUS, "sp.mode", self.MODES, "idle")
    self.final = circuit.get(PREVIOUS, "sp.final", (False, True), False).eq(True)
    self.quarter = circuit.get(PREVIOUS, "sp.quarter", tuple(range(4)), 0)

  def active(self):
    return NOT(OR(self.mode.eq("idle"), self.mode.eq("found"), self.mode.eq("missed")))

  def set_mode(self, value, enabled):
    self.mode = Value.select(enabled, Value.constant(value), self.mode)

  def start(self, lower, enabled=TRUE):
    self.circuit.require(AND(NOT(lower.negative()), NOT(self.radius.negative()),
                             NOT(self.center.read().eq(None))), enabled)
    self.program.reset(enabled)
    self.lower.copy_from(lower, enabled)
    self.work.copy_from(lower, enabled)
    self.work.inc(AND(enabled, self.work.zero()))
    self.span.reset(enabled)
    self.debt.pos.copy_from(self.radius.neg, enabled)
    self.debt.neg.copy_from(self.radius.pos, enabled)
    self.set_mode("grow", enabled)
    self.final = choose(enabled, FALSE, self.final)
    self.quarter = Value.select(enabled, Value.constant(0), self.quarter)

  def advance_match(self, enabled=TRUE):
    self.circuit.require(self.active(), enabled)
    self.radius.inc(enabled); self.debt.dec(enabled)

  def prepare(self, enabled):
    self.program.reset(enabled)
    self.walker.copy_from(self.center, enabled)
    self.work.copy_from(self.lower, enabled)
    tape = self.program.tapes[LOWER]
    tape.write(Value.constant(LEFT), enabled); tape.move(1, enabled)
    self.set_mode("lower", enabled)
    self.final = choose(enabled, FALSE, self.final)

  def double(self, enabled):
    self.work.copy_from(self.span, enabled)
    self.span.reset(enabled)
    self.quarter = Value.select(enabled, Value.constant(0), self.quarter)
    self.set_mode("double", enabled)

  def run_step(self, enabled):
    self.program.step(enabled)
    finished = AND(enabled, self.program.done)
    self.circuit.require(NOT(self.debt.negative()), finished)
    found = AND(finished, self.program.pc.eq(self.program.kernel.found))
    missed = AND(finished, NOT(found), self.final)
    next_stage = AND(finished, NOT(found), NOT(self.final), self.debt.zero())
    wait = AND(finished, NOT(found), NOT(self.final), NOT(self.debt.zero()))
    self.set_mode("found", found); self.set_mode("missed", missed)
    self.double(next_stage)
    self.set_mode("wait", wait)

  def step(self, enabled=TRUE):
    mode = {name: AND(enabled, self.mode.eq(name)) for name in self.MODES}
    source, lower = self.program.tapes[SOURCE], self.program.tapes[LOWER]
    positive = self.work.positive()
    grow = AND(mode["grow"], positive)
    self.work.dec(grow)
    for _ in range(8): self.span.inc(grow)
    for _ in range(2): self.debt.inc(grow)
    self.prepare(AND(mode["grow"], NOT(positive)))

    positive = self.work.positive()
    loading = AND(mode["lower"], positive)
    lower.write(Value.constant("1"), loading); lower.move(1, loading)
    self.work.dec(loading)
    loaded = AND(mode["lower"], NOT(positive))
    lower.write(Value.constant(END), loaded)
    self.set_mode("lower_home", loaded)

    home = lower.focus.eq(LEFT)
    begin_copy = AND(mode["lower_home"], home)
    source.write(Value.constant(LEFT), begin_copy); source.move(1, begin_copy)
    self.work.copy_from(self.span, begin_copy); self.work.inc(begin_copy)
    self.set_mode("copy", begin_copy)
    lower.move(-1, AND(mode["lower_home"], NOT(home)))

    symbol = self.walker.read()
    end = OR(symbol.eq(None), self.work.zero())
    stop_copy = AND(mode["copy"], end)
    source.write(Value.constant(END), stop_copy)
    self.final = choose(stop_copy, symbol.eq(None), self.final)
    self.set_mode("home", stop_copy)
    copying = AND(mode["copy"], NOT(end))
    source.write(Value({s: guard for s, guard in symbol.cases.items() if s is not None}), copying)
    source.move(1, copying); self.walker.left(copying); self.work.dec(copying)

    home = source.focus.eq(LEFT)
    start_run = AND(mode["home"], home)
    self.program.start(start_run); self.set_mode("run", start_run)
    source.move(-1, AND(mode["home"], NOT(home)))
    self.run_step(mode["run"])

    self.circuit.require(NOT(self.debt.negative()), mode["wait"])
    self.double(AND(mode["wait"], self.debt.zero()))

    positive = self.work.positive()
    doubling = AND(mode["double"], positive)
    quarter_end = self.quarter.eq(3)
    self.work.dec(doubling); self.span.inc(doubling); self.span.inc(doubling)
    self.quarter = Value.select(doubling, self.quarter.map(lambda n: (n + 1) % 4), self.quarter)
    self.debt.inc(AND(doubling, quarter_end))
    self.prepare(AND(mode["double"], NOT(positive)))

  def tick(self, enabled=TRUE, quantum=64):
    running = self.mode.eq("run")
    self.step(enabled)
    for _ in range(1, quantum):
      self.run_step(AND(enabled, running, self.mode.eq("run")))

  def finalize(self):
    self.program.finalize()
    for counter in (self.lower, self.span, self.work, self.debt): counter.finalize()
    self.circuit.put("sp.mode", self.mode)
    self.circuit.put("sp.final", Value.select(self.final, Value.constant(True), Value.constant(False)))
    self.circuit.put("sp.quarter", self.quarter)

File: scala/pal/src/main/scala/pal/ScaffoldCircuitSearch.scala
package pal

import Expr.{TRUE, FALSE}
import Ref.PREVIOUS
import ScaffoldCircuit.{conjunction as and, disjunction as or, neg as not, choose}
import ScaffoldCircuitInput.PlaceHead
import ScaffoldCircuitProgram.Program
import FppFinite.{LEFT, END}
import FppSubroutine.SOURCE
import DpFinite.LOWER

/** Predicated lowering of the match-paced DP search used by scaffold_galil. */
object ScaffoldCircuitSearch {
  val MODES: Vector[String] = Vector("idle", "grow", "lower", "lower_home", "copy", "home", "run",
    "found", "missed", "wait", "double")

  final class Search(val circuit: Circuit, val program: Program, pool: StackPool,
                     val center: PlaceHead, val walker: PlaceHead, val radius: Counter) {
    val lower: Counter = new Counter(pool, "sp.lo")
    val span: Counter = new Counter(pool, "sp.span")
    val work: Counter = new Counter(pool, "sp.work")
    val debt: Counter = new Counter(pool, "sp.debt")
    var mode: Value[String] = circuit.get(PREVIOUS, "sp.mode", MODES, "idle")
    var finalStage: Expr = circuit.get(PREVIOUS, "sp.final", Vector(false, true), false).eqTo(true)
    var quarter: Value[Int] = circuit.get(PREVIOUS, "sp.quarter", (0 until 4).toVector, 0)

    def active(): Expr = not(or(mode.eqTo("idle"), mode.eqTo("found"), mode.eqTo("missed")))
    def setMode(value: String, enabled: Expr): Unit = { mode = Value.select(enabled, Value.constant(value), mode) }

    def start(lowerBound: Counter, enabled: Expr = TRUE): Unit = {
      circuit.require(and(not(lowerBound.negative()), not(radius.negative()), not(center.read().eqTo(None))), enabled)
      program.reset(enabled)
      lower.copyFrom(lowerBound, enabled)
      work.copyFrom(lowerBound, enabled)
      work.inc(and(enabled, work.zero()))
      span.reset(enabled)
      debt.positiveStack.copyFrom(radius.negativeStack, enabled)
      debt.negativeStack.copyFrom(radius.positiveStack, enabled)
      setMode("grow", enabled)
      finalStage = choose(enabled, FALSE, finalStage)
      quarter = Value.select(enabled, Value.constant(0), quarter)
    }

    def advanceMatch(enabled: Expr = TRUE): Unit = {
      circuit.require(active(), enabled)
      radius.inc(enabled)
      debt.dec(enabled)
    }

    def prepare(enabled: Expr): Unit = {
      program.reset(enabled)
      walker.copyFrom(center, enabled)
      work.copyFrom(lower, enabled)
      val tape = program.tapes(LOWER)
      tape.write(Value.constant(LEFT), enabled)
      tape.move(1, enabled)
      setMode("lower", enabled)
      finalStage = choose(enabled, FALSE, finalStage)
    }

    def double(enabled: Expr): Unit = {
      work.copyFrom(span, enabled)
      span.reset(enabled)
      quarter = Value.select(enabled, Value.constant(0), quarter)
      setMode("double", enabled)
    }

    def runStep(enabled: Expr): Unit = {
      program.step(enabled)
      val finished = and(enabled, program.done)
      circuit.require(not(debt.negative()), finished)
      val found = and(finished, program.pc.eqTo(program.kernel.found.get))
      val missed = and(finished, not(found), finalStage)
      val nextStage = and(finished, not(found), not(finalStage), debt.zero())
      val waitForMatch = and(finished, not(found), not(finalStage), not(debt.zero()))
      setMode("found", found)
      setMode("missed", missed)
      double(nextStage)
      setMode("wait", waitForMatch)
    }

    def step(enabled: Expr = TRUE): Unit = {
      val dispatch = MODES.map(name => name -> and(enabled, mode.eqTo(name))).toMap
      val sourceTape = program.tapes(SOURCE)
      val lowerTape = program.tapes(LOWER)
      val growing = work.positive()
      val grow = and(dispatch("grow"), growing)
      work.dec(grow)
      for (_ <- 0 until 8) { span.inc(grow) }
      for (_ <- 0 until 2) { debt.inc(grow) }
      prepare(and(dispatch("grow"), not(growing)))

      val loadingPositive = work.positive()
      val loading = and(dispatch("lower"), loadingPositive)
      lowerTape.write(Value.constant("1"), loading)
      lowerTape.move(1, loading)
      work.dec(loading)
      val loaded = and(dispatch("lower"), not(loadingPositive))
      lowerTape.write(Value.constant(END), loaded)
      setMode("lower_home", loaded)

      val lowerHome = lowerTape.focus.eqTo(LEFT)
      val beginCopy = and(dispatch("lower_home"), lowerHome)
      sourceTape.write(Value.constant(LEFT), beginCopy)
      sourceTape.move(1, beginCopy)
      work.copyFrom(span, beginCopy)
      work.inc(beginCopy)
      setMode("copy", beginCopy)
      lowerTape.move(-1, and(dispatch("lower_home"), not(lowerHome)))

      val symbol = walker.read()
      val end = or(symbol.eqTo(None), work.zero())
      val stopCopy = and(dispatch("copy"), end)
      sourceTape.write(Value.constant(END), stopCopy)
      finalStage = choose(stopCopy, symbol.eqTo(None), finalStage)
      setMode("home", stopCopy)
      val copying = and(dispatch("copy"), not(end))
      sourceTape.write(Value(symbol.cases.collect { case (s, guard) if s != None => s.toString -> guard }), copying)
      sourceTape.move(1, copying)
      walker.left(copying)
      work.dec(copying)

      val sourceHome = sourceTape.focus.eqTo(LEFT)
      val startRun = and(dispatch("home"), sourceHome)
      program.start(startRun)
      setMode("run", startRun)
      sourceTape.move(-1, and(dispatch("home"), not(sourceHome)))
      runStep(dispatch("run"))

      circuit.require(not(debt.negative()), dispatch("wait"))
      double(and(dispatch("wait"), debt.zero()))

      val doublingPositive = work.positive()
      val doubling = and(dispatch("double"), doublingPositive)
      val quarterEnd = quarter.eqTo(3)
      work.dec(doubling)
      span.inc(doubling)
      span.inc(doubling)
      quarter = Value.select(doubling, quarter.map(n => (n + 1) % 4), quarter)
      debt.inc(and(doubling, quarterEnd))
      prepare(and(dispatch("double"), not(doublingPositive)))
    }

    def tick(enabled: Expr = TRUE, quantum: Int = 64): Unit = {
      val running = mode.eqTo("run")
      step(enabled)
      for (_ <- 1 until quantum) { runStep(and(enabled, running, mode.eqTo("run"))) }
    }

    def commit(): Unit = {
      program.commit()
      Vector(lower, span, work, debt).foreach(_.commit())
      circuit.put("sp.mode", mode)
      circuit.put("sp.final", Value.select(finalStage, Value.constant(true), Value.constant(false)))
      circuit.put("sp.quarter", quarter)
    }
  }
}
